docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,180 @@
 #!/usr/bin/env python3
 import argparse
+import hashlib
+import json
+import subprocess
+import typing
+
+
+class ImageNotFoundError(ValueError):
+    pass
+
+
+class _ImageKey(typing.NamedTuple):
+    commands_hash: str
+    packages_hash: str
+
+
+def _docker_push_latest_if_changed(source, target, *, is_dry_run):
+    _validate_source(source)
+    target = _get_sanitized_target(source, target)
+
+    print('Pushing source image')
+    _push_image(source, is_dry_run=is_dry_run)
+    try:
+        print('Pulling target image...')
+        _pull_image(target)
+    except ImageNotFoundError:
+        print((
+            f'Target image {target} was not found in the registry. '
+            'Going to attempt to tag and push the target image anyway.'
+        ))
+        _tag_image(source, target, is_dry_run=is_dry_run)
+        _push_image(target, is_dry_run=is_dry_run)
+    else:
+        if _has_image_changed(source, target):
+            print((
+                'Source image was found to be different from the current '
+                'version of the target image. Tagging and pushing new target '
+                'image.'
+            ))
+            _tag_image(source, target, is_dry_run=is_dry_run)
+            _push_image(target, is_dry_run=is_dry_run)
+        else:
+            print((
+                'Source image was found to be the same as the current '
+                'version of the target image. Not pushing the target image.'
+            ))
+
+
+def _validate_source(source):
+    print('Docker inspecting source to verify it exists')
+    source_inspect = _inspect_image(source)
```

Comment:
Since it won't let me reply to the old thread about just using urllib.parse for some reason :?: I'll put my thoughts here

Yeah, when using urllib.parse you'll need to prepend a dummy scheme so that the resolution works correctly (and then it should work like a charm!)

Something like:


```python
urllib.parse.urlparse(f'dummy://{source}')
```

Reply:
Oh yeah, that does work
Removing the `docker inspect` does mean I don't check if the image actually exists or not though.
We could just believe in the error message of the later `docker push` command instead, but then we'll end up getting false positive steps when `dry_run` is passed. What would finally throw an error for a not found image would be the `docker history` command when checking if the image has changed.

I'll probably keep an inspect image just to check for existence.